proposals/003-authentication-api.md

```diff
@@ -0,0 +1,223 @@
+Specifically:
+
+* If clients are authenticating to the proxy using mTLS then filters don't have access to a `Subject` or `Principal` corresponding to the client TLS certificate.
+* If clients are authenticating using the `SaslAuthenticate` KRPC the only way a `Filter` can know about that is by intercepting those frames.
+    - For "SASL passthrough", a `Filter` implementation could try to _infer_ the client's identity by watching for a successful `SaslAuthenticateResponse` 
+      returned by the broker. This is a technique that each identity-dependent filter in the chain could use, but each needs to reimplement the logic.
+    - For "SASL termination", a `Filter` implementation could handle the `SaslAuthenticateRequest` frames itself 
+        - If it makes its own `SaslAuthenticateRequests` (with different credentials, and/or a different mechanism) to the server then subsequent filters in the chain need to use the "SASL passthrough" technique to be aware of the broker-facing identity, and there's no way for them to know the original identity.
+        - If it doesn't (so that there is no `SaslAuthenticate` intersection with the broker) then subsequent filters in the chain have no way of knowing the client's identity.
+
+## Motivation
+
+The lack of this API makes implementing client identity aware plugins difficult, or impossible.
+
+Goals: 
+
+* Enable `Filters` to access a client's identity using a single, consistent API, irrespective of what authentication mechanism is implemented, TLS or SASL, and whether it's implemented by the proxy runtime (in the TLS case), or a prior Filter in the chain (in the SASL termination case).
+* Don't require a `Filter` to handle `SaslAuthenticate` unless it is performing SASL termination. 
+
+Non-goals:
+
+* Defining an API for exposing the identity of a _broker_ to `Filters` (in cases where the proxy mututally authenticates the broker).
+
+## Proposal
+
+### Public API changes
+
+Add the following interface to the `kroxylicious-api` module to allow a `Filter` implementation to opt into consuming authentication outcomes:
+
+```java
+package io.kroxylicious.proxy.filter;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+// This is a new interface in the api module
+interface ClientSubjectAwareFilter extends Filter {
+    // Called when a client authenticates, or reauthenticates
+    void onClientAuthentication(Subject clientSubject, FilterContext context);
+    // Called when a client fails authentication, or reauthentication
+    default void onClientAuthenticationFailure(LoginException exception, FilterContext context) { 
+    }
+}
+```
+
+Add the following two methods to `FilterContext` to allow SASL terminating `Filter` implementations to propagate their authentication outcomes to upstream filters:
+
+```java
+public interface FilterContext {
+
+    void clientAuthenticationSuccess(Subject subject);
```

Comment:
Would this synchronously invoke all the other filters in the chain with their onClientAuthentication*?

I'm wondering if it would help the Framework control the order-of-events if we make it something you can tack on to the result builder instead like:

`context.requestFilterResultBuilder().withClientAuthenticationSuccess(subject).forwardRequest(headers, response)`

This might make the users testing life easier, instead of having to check if some void method was invoked by the filter, they can assert the properties of the result object.

Also, it might help head off bugs (features?) like repeated calls to the auth methods, calling both success/fail methods, calling them at unexpected times from other threads etc.

Reply:
have discussed this with Tom and I think I'm worrying about some unlikely edge cases like:
1. Filters unexpectedly invoking the methods after returning a result
2. Filters unexpectedly invoking the methods multiple times etc

that can just be handled with good javadoc on the FilterContext methods describing how/when we expect the methods to be called. It's quite deep work/knowledge required to be working with auth anyway.